### PR TITLE
RFC: introduce a Non-conflicting insert operator

### PR DESCRIPTION
**Pull request by huitseeker**
This :
- demonstrates the use of a merge operator to implement a "weak insert" (insert on empty-or-equal).

This could:
- be used in a try-and-check logic with a get/multi-get to see if a batch of operations have all performed atomically,
- offer a limited substitute to an undefinable CAS operation on the DB.

This could not:
- implement an across-keys all-or-nothing operation, as any merge operator runs independent read-modify-write operations for each key.

TODO:
- [ ] make the operator opt-in rather than by-default,
- [ ] use `pre` crate to make user certify they installed the merge operator they expect at point of use,
- [ ] open the API to let users customize merge operators themselves: 
- ~demo multiplexing of merge operators~

### CODE REVIEW

File: typed-store/src/rocks/mod.rs
Comment (gdanezis) [inline typed-store/src/rocks/mod.rs]: I would use `Vec::extend` here? 
https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-Extend%3CT%3E

File: typed-store/src/rocks/mod.rs
Comment (gdanezis) [inline typed-store/src/rocks/mod.rs]: I think that nops represents the number of operands, not the length of the resulting vec. Maybe here it is worth allocating at least as much as `existing_val` if there is some of it?

File: typed-store/src/rocks/mod.rs
Comment (gdanezis) [inline typed-store/src/rocks/mod.rs]: I think this currently would write a long string which is the concatenation of all the merges. Is this what is intended here? I would have expected you want to write the very first merge, and then ignore the remaining ones? If so I would change this line to something like
`if let Some(op) = operands.next()`
and then ignore the remaining ones.

You can make a test case that does 2 merges then a read to check if my understanding above is correct.

File: typed-store/src/rocks/mod.rs
Comment (gdanezis) [inline typed-store/src/rocks/mod.rs:474]: I would allow users of the DB to set this optionally, particularly if there is some magic typing trick to only expose the merge based operations for DBs that have this option on.

**Review by gdanezis — COMMENTED** [review]
Thanks for looking into merge ops -- this will become crucial down the line. I have a few comments inline, incl one that needs confirmation on the semantics of multiple merges to a non existstant key.

In terms of the motivation stated: how can use this as a substitute for the Read-Modify-Write operations? The issue I see with this is that such RMW sequence involves a read, which would immediately flatten the merge, therefore not getting much benefit from it.

I think the mechanism we could leverage instead is the GetForUpdate in upstream rocksdb:
https://github.com/facebook/rocksdb/wiki/Transactions#guarding-against-read-write-conflicts
which I have not yet seen exposed in the rust wrapper :(

File: typed-store/src/rocks/mod.rs
Comment (huitseeker) [reply typed-store/src/rocks/mod.rs]: Your understanding is correct, but the "long string" behavior only happens in highly concurrent cases, as the tests (passing before the "fix") show.

**Comment by huitseeker** [comment]
> In terms of the motivation stated:

Essentially:
```rust
while db.batch().multi_get(keys_n_vals.map(|(k, v)| k).unwrap() != keys_n_vals.map(|(k, v)| Some(v)).collect::<Vec<_>>() {
    db.batch().non_conflicting_insert_batch(&db, keys_n_vals).write();
}
<do something requiring a weak multi-key lock>
```

Weak here means not pre-empted by another non-conflicting insert op.

It's a poor substitute for an in-memory lock and would likely need to work in conjunction with one of these.

> how can use this as a substitute for the Read-Modify-Write operations? 

I don't understand the question. This *is* a specific RMW op, it does not aim to offer a substitute for the generality of the class of these operations.

> I think the mechanism we could leverage instead is the GetForUpdate in upstream rocksdb:

This is essentially blocked on https://github.com/rust-rocksdb/rust-rocksdb/pull/250

File: typed-store/src/rocks/mod.rs
Comment (huitseeker) [reply typed-store/src/rocks/mod.rs:474]: Probable. It's in the TODO and will be dealt with before merge.

**Review by gdanezis — APPROVED** [review]
Looks good!

File: typed-store/src/rocks/mod.rs
Comment (velvia) [inline typed-store/src/rocks/mod.rs:459]: There's no need to compute nops and allocate a result if there is an existing val, it's wasted.  Also not sure how fast v.iter() is, but it's probably optimized.  However I would do this which should be faster and simpler:

```rust
if let Some(v) = existing_val {
    Some(v.clone())
} else if let Some(op) = operands.into_iter().next() {
    Some(op.clone())
}
```

File: typed-store/src/rocks/tests.rs
Comment (velvia) [inline typed-store/src/rocks/tests.rs:117]: Maybe this is just me but I wish there was a way to return a value indicating whether the write succeeded or not.  To return OK when a write didn't happen seems to be, non ideal from API perspective.

**Review by velvia — COMMENTED** [review]
Neat.
I can't tell from the RocksDB docs but it seems like you could pass in different merge operator with each C++ API merge, but here in the Rust API you have to set a single merge operator for the entire CF which is too bad.....

File: typed-store/src/rocks/mod.rs
Comment (huitseeker) [reply typed-store/src/rocks/mod.rs:459]: This is not a total map (`existing_val` and `operands` can both be `None`), and `v`,  `op` are slices, which require a `to_vec()` to allocate besides. But I took some of the feedback into account nonetheless.

**Pull request closed without merging**
